packages/filesystem/src/browser/file-tree/file-tree-widget.ts

```diff
@@ -56,4 +63,108 @@ export class FileTreeWidget extends TreeWidget {
+            this.addEventListener(this.node, 'dragover', event => this.handleDragOverEvent(treeNode, event));
+            this.addEventListener(this.node, 'dragleave', event => this.handleLeaveEvent(event));
+            this.addEventListener(this.node, 'drop', event => this.handleDropEvent(treeNode, event, false));
+        }
+    }
+
+    protected createNodeAttributes(node: ITreeNode, props: NodeProps): ElementAttrs {
+        const elementAttrs: ElementAttrs = super.createNodeAttributes(node, props);
+        const childExpansion = this.isExpandable(node) ? 0 : 1;
+
+        if (!childExpansion) {
+            // Folder, so it could be a dropzone
+            return {
+                ...elementAttrs,
+                ondrop: event => this.handleDropEvent(node, event, true),
+            };
+        }
+        return elementAttrs;
+    }
+
+    protected handleDragEnterEvent(node: ITreeNode, event: DragEvent): void {
+        event.preventDefault();
+        event.stopPropagation();
+        const target = event.target as HTMLElement;
+        if (target.parentElement) {
+            target.parentElement.classList.add(activeDropZone);
+        }
+    }
+
+    protected handleDragOverEvent(node: ITreeNode, event: DragEvent): void {
+        event.preventDefault();
+    }
+
+    protected handleLeaveEvent(event: DragEvent): void {
+        event.preventDefault();
+        event.stopPropagation();
+        const target = event.target as HTMLElement;
+        if (target.parentElement) {
+            target.parentElement.classList.remove(activeDropZone);
+        }
+    }
+
+    protected handleDropEvent(node: ITreeNode, event: DragEvent, dropAllowed: boolean): void {
+        event.stopPropagation();
+        event.preventDefault();
+        if (dropAllowed) {
+            event.dataTransfer.dropEffect = 'copy'; // Explicitly show this is a copy.
+            const target = event.target as HTMLElement;
+            if (target.parentElement) {
+                target.parentElement.classList.remove(activeDropZone);
+            }
+            this.dropHandler(event, node);
+        }
+    }
+
+    protected dropHandler(ev: DragEvent, node: ITreeNode) {
+        ev.preventDefault();
+        const data = ev.dataTransfer.items;
+
+        for (let i = 0; i < data.length; i += 1) {
+            if (data[i].kind === 'file') {
+                const entry: WebKitEntry = data[i].webkitGetAsEntry();
+
+                if (entry.isDirectory) {
+                    this.handleDirectory(entry, node.id);
+                } else {
+                    const f: any = data[i].getAsFile();
+                    this.model.upload(node.id, f, f.name);
+                }
+            }
+        }
+    }
+
+    protected handleDirectory(webDirEntry: WebKitEntry, dest: string) {
+
+        const getEntries = (dirEntry: any, destination: string) => {
+            const reader = dirEntry.createReader();
+            reader.readEntries((e: any) => {
```

Comment:
`(e: WebKitEntry[])` instead of downcasting in the closure body

Reply:
OK, done